lm-agent/lm_agent/parsing/flexlm.py

```diff
@@ -11,63 +12,184 @@
+    rf"\((?P<license_host>{HOSTNAME})/"
+    rf"(?P<license_port>{INT}) "
+    rf"(?P<license_pid>{INT})\), "
+    r"start "
+    rf"(?P<weekday>{NONWS}) "
+    rf"(?P<date>{MMDD}) "
+    rf"(?P<time>{HHMM}), "
+    rf"(?P<tokens>{INT}) "
+    r"licenses$"
 )
 
 
-RX = re.compile(rf"{DATA_LINE}|{TOTALS_LINE}")
+USAGE_LINE_3 = (
+    r"^\s+"
+    rf"(?P<user>{NONWS}) "
+    rf"(?P<user_host>{HOSTNAME}) "
+    rf"(?P<tty>{NONWS}) "
+    rf"\(v(?P<version>{NONPAREN})\) "
+    rf"\((?P<license_host>{HOSTNAME})/"
+    rf"(?P<license_port>{INT}) "
+    rf"(?P<license_pid>{INT})\), "
+    r"start "
+    rf"(?P<weekday>{NONWS}) "
+    rf"(?P<date>{MMDD}) "
+    rf"(?P<time>{HHMM})$"
+)
 
+USAGE_LINE_4 = (
+    r"^\s+"
+    rf"(?P<user>{NONWS}) "
+    rf"(?P<user_host>{HOSTNAME}) "
+    rf"(?P<tty>{HOSTNAME}) "
+    rf"(?P<feature>{FEATURE_NAME}) "
+    rf"\(v(?P<version>{NONPAREN})\) "
+    rf"\((?P<license_host>{HOSTNAME})/"
+    rf"(?P<license_port>{INT}) "
+    rf"(?P<license_pid>{INT})\), "
+    r"start "
+    rf"(?P<weekday>{NONWS}) "
+    rf"(?P<date>{MMDD}) "
+    rf"(?P<time>{HHMM}), "
+    rf"(?P<tokens>{INT}) "
+    r"licenses$"
+)
+
+USAGE_LINE_5 = (
+    r"^\s+"
+    rf"(?P<user>{NONWS}) "
+    rf"(?P<user_host>{HOSTNAME}) "
+    rf"(?P<tty>{NONWS}) "
+    rf"(?P<feature>{FEATURE_NAME}) "
+    rf"\(v(?P<version>{NONPAREN})\) "
+    rf"\((?P<license_host>{HOSTNAME})/"
+    rf"(?P<license_port>{INT}) "
+    rf"(?P<license_pid>{INT})\), "
+    r"start "
+    rf"(?P<weekday>{NONWS}) "
+    rf"(?P<date>{MMDD}) "
+    rf"(?P<time>{HHMM})$"
+)
+
+RX_FEATURE = re.compile(FEATURE_LINE)
+RX_USAGE_LINE_1 = re.compile(USAGE_LINE_1)
+RX_USAGE_LINE_2 = re.compile(USAGE_LINE_2)
+RX_USAGE_LINE_3 = re.compile(USAGE_LINE_3)
+RX_USAGE_LINE_4 = re.compile(USAGE_LINE_4)
+RX_USAGE_LINE_5 = re.compile(USAGE_LINE_5)
+USAGE_LINES = [RX_USAGE_LINE_1, RX_USAGE_LINE_2, RX_USAGE_LINE_3, RX_USAGE_LINE_4, RX_USAGE_LINE_5]
+
+
+def parse_feature_line(line: str) -> Optional[Dict]:
+    """
+    Parse the feature line in the FlexLM output.
+    Data we need:
+    - ``feature``: license name
+    - ``total``: total amount of licenses
+    - ``used``: quantity of licenses being use
+    """
+    parsed_feature = RX_FEATURE.match(line)
+    if parsed_feature is None:
+        return None
+    feature_data = parsed_feature.groupdict()
+
+    return {
+        "feature": feature_data["feature"].lower(),
+        "total": int(feature_data["total"]),
+        "used": int(feature_data["used"]),
+    }
+
+
+def parse_usage_line(line: str) -> Optional[Dict]:
+    """
+    Parse the usage line in the FlexLM output.
+    Data we need:
+    - ``user_name``: user who booked the license
+    - ``lead_host``: host using the license
+    - ``booked``: quantity of licenses being used
+
+    There can be multiple formats for the data line, so we need to check which one matches.
+    """
+
+    for RX in USAGE_LINES:
+        parsed_data = RX.match(line)
+        if parsed_data:
```

Comment:
Opportunity for walrus operator!

```suggestion
        if parsed_data := RX.match(line):
```

Reply:
First time using it, nice!